File: apps/website/src/oidc.ts
export function toWebSocketUrl(baseUrl: string): string {
  const url = new URL(baseUrl);
  url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
  url.pathname = "/ws";
  url.search = "";
  return url.toString();
}

function base64UrlEncode(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}

export function createCodeVerifier(byteLength = 32): string {
  const bytes = new Uint8Array(byteLength);
  crypto.getRandomValues(bytes);
  return base64UrlEncode(bytes);
}

export async function createCodeChallenge(verifier: string): Promise<string> {
  const bytes = new TextEncoder().encode(verifier);
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  return base64UrlEncode(new Uint8Array(digest));
}

interface AuthConfig {
  clientId: string;
  authorizationEndpoint: string;
  tokenEndpoint: string;
}

interface OidcState {
  state: string;
  codeVerifier: string;
  backendUrl: string;
  redirectUri: string;
}

const OIDC_STATE_KEY = "vxbeamer_oidc_state";

async function fetchAuthConfig(backendUrl: string): Promise<AuthConfig> {
  const res = await fetch(new URL("/auth/config", backendUrl).toString());
  if (!res.ok) throw new Error("Failed to fetch auth config");
  return res.json() as Promise<AuthConfig>;
}

export async function startSignIn(backendUrl: string): Promise<never> {
  const config = await fetchAuthConfig(backendUrl);

  const codeVerifier = createCodeVerifier();
  const codeChallenge = await createCodeChallenge(codeVerifier);
  const state = crypto.randomUUID();
  const redirectUri = window.location.origin + window.location.pathname;

  const oidcState: OidcState = { state, codeVerifier, backendUrl, redirectUri };
  sessionStorage.setItem(OIDC_STATE_KEY, JSON.stringify(oidcState));

  const authUrl = new URL(config.authorizationEndpoint);
  authUrl.searchParams.set("response_type", "code");
  authUrl.searchParams.set("client_id", config.clientId);
  authUrl.searchParams.set("redirect_uri", redirectUri);
  authUrl.searchParams.set("scope", "openid profile");
  authUrl.searchParams.set("code_challenge", codeChallenge);
  authUrl.searchParams.set("code_challenge_method", "S256");
  authUrl.searchParams.set("state", state);

  window.location.href = authUrl.toString();
  throw new Error("Redirecting");
}

type CallbackResult =
  | { type: "session"; accessToken: string; refreshToken: string; backendUrl: string }
  | { type: "desktop"; payload: string };

export async function createAuthUrl(backendUrl: string): Promise<{
  url: string;
  codeVerifier: string;
  state: string;
}> {
  const config = await fetchAuthConfig(backendUrl);

  const codeVerifier = createCodeVerifier();
  const codeChallenge = await createCodeChallenge(codeVerifier);
  const state = "desktop:" + crypto.randomUUID();

  const authUrl = new URL(config.authorizationEndpoint);
  authUrl.searchParams.set("response_type", "code");
  authUrl.searchParams.set("client_id", config.clientId);
  authUrl.searchParams.set("redirect_uri", window.location.origin + window.location.pathname);
  authUrl.searchParams.set("scope", "openid profile");
  authUrl.searchParams.set("code_challenge", codeChallenge);
  authUrl.searchParams.set("code_challenge_method", "S256");
  authUrl.searchParams.set("state", state);

  return {
    url: authUrl.toString(),
    codeVerifier,
    state,
  };
}

export async function exchangeDesktopCode(
  payload: string,
  codeVerifier: string,
  expectedState: string,
  backendUrl: string,
): Promise<{ accessToken: string; refreshToken: string }> {
  const [code, returnedState] = payload.split("#", 2) as [string, string];

  if (!code || !returnedState) {
    throw new Error("Invalid code format — expected <code>#<state>");
  }

  if (returnedState !== expectedState) {
    throw new Error("State mismatch");
  }

  const redirectUri = window.location.origin + window.location.pathname;
  const config = await fetchAuthConfig(backendUrl);

  const tokenRes = await fetch(config.tokenEndpoint, {
    method: "POST",
    headers: { "content-type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "authorization_code",
      client_id: config.clientId,
      code,
      redirect_uri: redirectUri,
      code_verifier: codeVerifier,
    }),
  });
  if (!tokenRes.ok) throw new Error("Token exchange failed");

  const tokens = (await tokenRes.json()) as { id_token?: string };
  if (!tokens.id_token) throw new Error("No id_token in token response");

  const sessionRes = await fetch(new URL("/auth/session", backendUrl).toString(), {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ id_token: tokens.id_token }),
  });
  if (!sessionRes.ok) {
    const err = (await sessionRes.json()) as { error?: string };
    throw new Error(err.error ?? "Session creation failed");
  }

  const session = (await sessionRes.json()) as { access_token?: string; refresh_token?: string };
  if (!session.access_token) throw new Error("No access_token in session response");
  if (!session.refresh_token) throw new Error("No refresh_token in session response");

  return { accessToken: session.access_token, refreshToken: session.refresh_token };
}

export async function handleCallback(): Promise<CallbackResult | null> {
  const params = new URLSearchParams(window.location.search);
  const code = params.get("code");
  const returnedState = params.get("state");

  if (!code || !returnedState) return null;

  // Check for desktop flow: state starts with "desktop:" and no sessionStorage entry
  const stored = sessionStorage.getItem(OIDC_STATE_KEY);
  if (returnedState.startsWith("desktop:") && !stored) {
    window.history.replaceState({}, "", window.location.pathname);
    return { type: "desktop", payload: `${code}#${returnedState}` };
  }

  if (!stored) throw new Error("No OIDC state found — please sign in again");

  const { state, codeVerifier, backendUrl, redirectUri } = JSON.parse(stored) as OidcState;
  sessionStorage.removeItem(OIDC_STATE_KEY);

  if (returnedState !== state) throw new Error("State mismatch");

  window.history.replaceState({}, "", window.location.pathname);

  const config = await fetchAuthConfig(backendUrl);

  const tokenRes = await fetch(config.tokenEndpoint, {
    method: "POST",
    headers: { "content-type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "authorization_code",
      client_id: config.clientId,
      code,
      redirect_uri: redirectUri,
      code_verifier: codeVerifier,
    }),
  });
  if (!tokenRes.ok) throw new Error("Token exchange failed");

  const tokens = (await tokenRes.json()) as { id_token?: string };
  if (!tokens.id_token) throw new Error("No id_token in token response");

  const sessionRes = await fetch(new URL("/auth/session", backendUrl).toString(), {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ id_token: tokens.id_token }),
  });
  if (!sessionRes.ok) {
    const err = (await sessionRes.json()) as { error?: string };
    throw new Error(err.error ?? "Session creation failed");
  }

  const session = (await sessionRes.json()) as { access_token?: string; refresh_token?: string };
  if (!session.access_token) throw new Error("No access_token in session response");
  if (!session.refresh_token) throw new Error("No refresh_token in session response");

  return {
    type: "session",
    accessToken: session.access_token,
    refreshToken: session.refresh_token,
    backendUrl,
  };
}

File: apps/watch/mobile/src/main/java/com/dtinth/vxbeamer/mobile/Oidc.kt
package com.dtinth.vxbeamer.mobile

import android.util.Base64
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.UUID
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/**
 * Same "desktop" sign-in flow the Tauri app already uses (dtinth/vxbeamer#86)
 * — this app never receives the OIDC redirect itself. The browser lands on
 * vxbeamer's hosted web app, which shows a "here is your code" screen (its
 * `redirect_uri` is fixed to that page, see [DESKTOP_REDIRECT_URI]); the
 * operator copies that code and pastes it in here. This mirrors
 * `apps/website/src/oidc.ts`'s `createAuthUrl`/`exchangeDesktopCode`
 * function for function — nothing on the backend or website needed to
 * change for a second app to reuse this same path.
 */
object Oidc {
    private const val DESKTOP_REDIRECT_URI = "https://vxbeamer.vercel.app/"
    private val client = OkHttpClient()

    data class AuthConfig(val clientId: String, val authorizationEndpoint: String, val tokenEndpoint: String)

    data class PendingSignIn(val authUrl: String, val codeVerifier: String, val state: String)

    data class Tokens(val accessToken: String, val refreshToken: String)

    suspend fun fetchAuthConfig(backendUrl: String): AuthConfig =
        onIoDispatcher {
            val request = Request.Builder().url(urlJoin(backendUrl, "/auth/config")).build()
            val body = executeForBody(request, "Failed to fetch auth config")
            val json = JSONObject(body)
            AuthConfig(
                clientId = json.getString("clientId"),
                authorizationEndpoint = json.getString("authorizationEndpoint"),
                tokenEndpoint = json.getString("tokenEndpoint"),
            )
        }

    suspend fun beginSignIn(backendUrl: String): PendingSignIn {
        val config = fetchAuthConfig(backendUrl)
        val codeVerifier = randomUrlSafeString(32)
        val codeChallenge = sha256UrlSafe(codeVerifier)
        val state = "desktop:${UUID.randomUUID()}"

        val authUrl =
            config.authorizationEndpoint.toHttpUrl().newBuilder()
                .addQueryParameter("response_type", "code")
                .addQueryParameter("client_id", config.clientId)
                .addQueryParameter("redirect_uri", DESKTOP_REDIRECT_URI)
                .addQueryParameter("scope", "openid profile")
                .addQueryParameter("code_challenge", codeChallenge)
                .addQueryParameter("code_challenge_method", "S256")
                .addQueryParameter("state", state)
                .build()

        return PendingSignIn(authUrl.toString(), codeVerifier, state)
    }

    /** [pastedCode] is the `<code>#<state>` string copied from the browser. */
    suspend fun completeSignIn(
        backendUrl: String,
        pastedCode: String,
        codeVerifier: String,
        expectedState: String,
    ): Tokens =
        onIoDispatcher {
            val parts = pastedCode.split("#", limit = 2)
            require(parts.size == 2) { "Invalid code format — expected <code>#<state>" }
            val (code, returnedState) = parts
            require(returnedState == expectedState) { "State mismatch — did you paste the right code?" }

            val config = fetchAuthConfig(backendUrl)
            val tokenBody =
                FormBody.Builder()
                    .add("grant_type", "authorization_code")
                    .add("client_id", config.clientId)
                    .add("code", code)
                    .add("redirect_uri", DESKTOP_REDIRECT_URI)
                    .add("code_verifier", codeVerifier)
                    .build()
            val tokenRequest = Request.Builder().url(config.tokenEndpoint).post(tokenBody).build()
            val tokenResponseBody = executeForBody(tokenRequest, "Token exchange failed")
            val idToken =
                JSONObject(tokenResponseBody).optString("id_token").ifEmpty {
                    error("No id_token in token response")
                }

            val sessionRequest =
                Request.Builder()
                    .url(urlJoin(backendUrl, "/auth/session"))
                    .post(JSONObject().put("id_token", idToken).toString().toRequestBody(JSON))
                    .build()
            val sessionBody = executeForBody(sessionRequest, "Session creation failed")
            val session = JSONObject(sessionBody)
            Tokens(
                accessToken = session.getString("access_token"),
                refreshToken = session.getString("refresh_token"),
            )
        }

    suspend fun refresh(backendUrl: String, refreshToken: String): Tokens =
        onIoDispatcher {
            val request =
                Request.Builder()
                    .url(urlJoin(backendUrl, "/auth/refresh"))
                    .post(JSONObject().put("refresh_token", refreshToken).toString().toRequestBody(JSON))
                    .build()
            val body = executeForBody(request, "Refresh failed")
            val session = JSONObject(body)
            Tokens(
                accessToken = session.getString("access_token"),
                refreshToken = session.optString("refresh_token", refreshToken),
            )
        }

    private fun executeForBody(request: Request, errorPrefix: String): String {
        client.newCall(request).execute().use { response ->
            val body = response.body?.string().orEmpty()
            if (!response.isSuccessful) error("$errorPrefix: HTTP ${response.code} $body")
            return body
        }
    }

    private val JSON = "application/json".toMediaType()

    private suspend fun <T> onIoDispatcher(block: suspend () -> T): T = withContext(Dispatchers.IO) { block() }

    private fun randomUrlSafeString(byteLength: Int): String {
        val bytes = ByteArray(byteLength)
        SecureRandom().nextBytes(bytes)
        return Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
    }

    private fun sha256UrlSafe(input: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(input.toByteArray(Charsets.UTF_8))
        return Base64.encodeToString(digest, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
    }

    private fun urlJoin(base: String, path: String): String = base.trimEnd('/') + path
}
